lib/oauth2-client.js

```diff
@@ -119,13 +135,52 @@ async function _requestAccessToken({url, body, httpsAgent}) {
     }));
   } catch(error) {
     logger.error('Error getting access token.', {error});
-    if(isUnrecoverableError({error})) {
-      logger.error('An unrecoverable error was encountered, exiting.');
-      bedrock.exit();
-    }
   }
   if(response && response.access_token) {
     return response.access_token;
   }
+
   return false;
 }
+
+async function getAccessTokenOnInit({
+  client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent
+}) {
+  const body = new URLSearchParams({
+    client_id, client_secret, grant_type, scope
+  });
+  let accessToken;
+  try {
+    accessToken = await pRetry(() => _requestAccessTokenOnInit({
+      url: token_endpoint, body, httpsAgent
+    }), {
+      retries: 300,
+      onFailedAttempt: async error => {
+        logger.error(
+          'Unable to acquire access token. Retrying...', {error});
+      },
+      maxTimeout: RETRY_MAX_TIMEOUT,
+    });
+  } catch(e) {
+    logger.debug('Failed to acquire access token.', {e});
+    throw e;
+  }
+  return accessToken;
+}
+
+async function _requestAccessTokenOnInit({url, body, httpsAgent}) {
+  let response;
+  try {
+    ({data: response} = await httpClient.post(url, {
+      body,
+      agent: httpsAgent
+    }));
+    return response.access_token;
+  } catch(error) {
+    if(isUnrecoverableError({error})) {
+      logger.error('An unrecoverable error was encountered, exiting.', {error});
+      bedrock.exit();
```

Comment:
It seems odd to me to put the exit in the client, but I understand it's being done here in one place so that it doesn't have to be repeated for all users of this module. If this is really what is being controlled by the `onInit` flag, then perhaps it should be called `exitOnFailure` or something like that.

Reply:
Thinking on this more, why not have this module listen for `bedrock.ready` itself and refresh its own access token -- and exit on that event if it fails? Do we need the flag? If we're concerned about there being multiple services for the client to connect to and ready itself, why not have a config for this module that allows multiple services to be setup? It seems that we're just going to duplicate the code to refresh the access token everywhere (just like we'd have to duplicate the exit code).

I now suggest we eliminate the special flag and have this module handle `bedrock.ready` itself and obtain the access token(s) it needs and exit if it fails to do so.

@dmitrizagidulin, @mandyvenables, thoughts? Is there a major drawback to this approach?

Reply:
I renamed it to `retryAndExitOnFailure`, I think that is the most descriptive while still being concise.

@dlongley I added in the flag because we only want the behavior (unlimited retries and exiting if we can't get the token) on start up; afterwards we still need to refresh the access token at set intervals, and we don't want the unlimited retry behavior  because then we have other calls waiting on the response.

Reply:
> I added in the flag because we only want the behavior (unlimited retries and exiting if we can't get the token) on start up...

Right, but that's a use case, not a description of the feature. There may be other use cases in the future. If we're convinced that that is unlikely, we should consider if other designs would be better.

To that end, it seems that what we're really trying to do here is avoid duplicating error handling code in numerous places by pushing it into this library. However, we're still adding event handling code everywhere to call into this library. Therefore, I'm still thinking we want this library to handle obtaining its initial access token on startup on its own -- and exit on its own -- removing the need for the flag entirely.

Why don't we put the `bedrock.ready` handler code into this module and allow this module to be configured with N many target servers (for which it will obtain N many access tokens on startup)?

Reply:
@dlongley - I agree that's a more preferable design (the library to handle obtaining the initial access token on startup on its own). 

I do want to make sure we do it in stages, though -- let's get this PR in (since it's addressing having to do manual token updates every 2 weeks), and iterate on the design.